pkg/k8s/watchers/cilium_node.go

```diff
@@ -5,153 +5,159 @@ package watchers
+			apiGroup,
 		)
-		isConnected := make(chan struct{})
-		// once isConnected is closed, it will stop waiting on caches to be
-		// synchronized.
-		k.blockWaitGroupToSyncResources(isConnected, swgNodes, ciliumNodeInformer.HasSynced, apiGroup)
-
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = ciliumNodeStore
-		k.ciliumNodeStoreMU.Unlock()
-
-		once.Do(func() {
-			// Signalize that we have put node controller in the wait group
-			// to sync resources.
-			asyncControllers.Done()
-		})
 		k.k8sAPIGroups.AddAPI(apiGroup)
-		go ciliumNodeInformer.Run(isConnected)
 
-		<-kvstore.Connected()
-		log.Info("Connected to key-value store, stopping CiliumNode watcher")
+		// Signalize that we have put node controller in the wait group to sync resources.
+		once.Do(asyncControllers.Done)
+
+		// Set forceCiliumNodeFromApiServer to false so that any attempts of getting the
+		// CiliumNode go through the Resource[T] cache.
+		k.forceCiliumNodeFromApiServer.Store(false)
 
-		// Set the ciliumNodeStore as nil so that any attempts of getting the
-		// CiliumNode are performed with a request sent to kube-apiserver
-		// directly instead of relying on an outdated version of the CiliumNode
-		// in this cache.
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = nil
-		k.ciliumNodeStoreMU.Unlock()
+		go func() {
+			// derive another context to signal Events() in case of kvstore connection
+			ctx, cancel := context.WithCancel(ctx)
+
+			defer func() {
+				cancel()
+				close(stop)
+			}()
+
+			events := k.resources.CiliumNode.Events(ctx)
+			cache := make(map[resource.Key]*cilium_v2.CiliumNode)
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-kvstore.Connected():
+					log.Info("Connected to key-value store, stopping CiliumNode watcher")
+					// Set forceCiliumNodeFromApiServer to true so that any attempts of getting the
+					// CiliumNode are performed with a request sent to kube-apiserver directly instead
+					// of relying on an outdated version of the CiliumNode in the Resource[T] cache.
+					k.forceCiliumNodeFromApiServer.Store(true)
+					k.cancelWaitGroupToSyncResources(apiGroup)
+					k.k8sAPIGroups.RemoveAPI(apiGroup)
+					return
+				case event, ok := <-events:
+					if !ok {
+						return
+					}
+					var err error
+					switch event.Kind {
+					case resource.Sync:
+						synced.Store(true)
+					case resource.Upsert:
+						var needUpdate bool
+						oldObj, ok := cache[event.Key]
+						if !ok {
+							needUpdate = k.onCiliumNodeInsert(event.Object)
+						} else {
+							needUpdate = k.onCiliumNodeUpdate(oldObj, event.Object)
+						}
+						if needUpdate {
+							cache[event.Key] = event.Object.DeepCopy()
+						}
+					case resource.Delete:
+						k.onCiliumNodeDelete(event.Object)
+						delete(cache, event.Key)
+					}
+					event.Done(err)
+				}
+			}
+		}()
 
-		close(isConnected)
+		select {
+		case <-ctx.Done():
+			return
+		case <-kvstore.Client().Disconnected():
```

Comment:
How is the goroutine spawned above stopped in this case? I'm guessing there is some interaction with Resource[T] but maybe it makes sense to document that.

Maybe it would be easier to understand what is going on if the `<-kvstore.Connected()` part also happened here?  Bascially `cancel` on the context of the goroutine reading from the event stream.